wrappers/wrapper.rb

```diff
@@ -742,89 +730,63 @@ def detect_unfeasible_services(vrp)
 
-    def check_distances(vrp, unfeasible)
-      unfeasible = check(vrp, :time, unfeasible)
-      unfeasible = check(vrp, :distance, unfeasible)
-      unfeasible = check(vrp, :value, unfeasible)
+    def service_reachable_by_vehicle_within_timewindows(vrp, activity, vehicle)
+      vehicle_start = vehicle.timewindow&.start || vehicle.sequence_timewindows.collect(&:start).min || 0
+      vehicle_end = vehicle.timewindow&.end || vehicle.sequence_timewindows.collect(&:end).max
+      matrix = vrp.matrices.find{ |m| m.id == vehicle.matrix_id }
 
-      # check distances from vehicle depot is feasible
-      vrp.services.each{ |service|
-        # Check via vehicle time-windows
-        reachable_by_a_vehicle = (service.activity ? [service.activity] : service.activities).any?{ |activity|
-          index = activity.point.matrix_index
-          vrp.vehicles.find{ |vehicle|
-            feasible = true
-
-            start_index = vehicle.start_point&.matrix_index
-            end_index = vehicle.end_point&.matrix_index
-            matrix = vrp.matrices.find{ |mat| mat.id == vehicle.matrix_id } if start_index || end_index
+      time_to_go = vehicle.start_point&.matrix_index ? matrix.time[vehicle.start_point&.matrix_index][activity.point.matrix_index] : 0
+      time_back = vehicle.end_point&.matrix_index ? matrix.time[activity.point.matrix_index][vehicle.end_point&.matrix_index] : 0
 
-            if vehicle.cost_time_multiplier&.positive? &&
-               !vehicle.cost_late_multiplier&.positive? &&
-               ((vehicle.timewindow&.start && vehicle.timewindow&.end) || vehicle.sequence_timewindows&.size&.positive?)
-              vehicle_available_time = vehicle.timewindow.end - vehicle.timewindow.start if vehicle.timewindow
-              vehicle_available_time = vehicle.sequence_timewindows.collect{ |tw| tw.end - tw.start }.max if vehicle.sequence_timewindows&.size&.positive?
+      earliest_arrival = vehicle_start + time_to_go
+      earliest_back = earliest_arrival + activity.duration + time_back
+      latest_arrival = vehicle_end - time_back - activity.duration if vehicle_end
 
-              min_time = (start_index && matrix[:time][start_index][index]).to_f + (end_index && matrix[:time][index][end_index]).to_f + activity.duration
+      if !activity.late_multiplier&.positive? # if service lateness is not allowed
+        return false if activity.timewindows.none?{ |tw| tw.end.nil? || earliest_arrival <= tw.end }
+      end
+      
+      return false if latest_arrival && activity.timewindows.all?{ |tw| latest_arrival < tw.start }
 
-              feasible &&= (vehicle_available_time >= min_time)
-            end
+      unless vehicle.cost_late_multiplier&.positive? && vehicle.cost_time_multiplier&.positive? &&
+             (vehicle_end || vehicle.duration)
+        return false if vehicle_end && earliest_back > vehicle_end
+        return false if vehicle.duration && earliest_back - earliest_arrival > vehicle.duration
+      end
```

Comment:
This `unless` statement looks incorrect to me, it should be something like this
```suggestion
      if vehicle.cost_time_multiplier&.positive? && !vehicle.cost_late_multiplier&.positive? # lateness is not allowed
        return false if vehicle_end && earliest_back > vehicle_end
        return false if vehicle.duration && earliest_back - earliest_arrival > vehicle.duration
      end
```